File: src/services/wiki/health.ts
import { resolveWikiRoot } from "./wiki-root.js";
import {
  collectHealthReport as collectIndexerHealthReport,
  DEFAULT_STALE_SECONDS,
  type HealthReport,
} from "../../stores/wiki/health/index.js";

export interface CheckWikiHealthRequest {
  cwd: string;
  wiki?: string;
  topic?: string;
  staleSeconds: number;
  stdinSlugs?: string[];
}

export interface CollectWikiHealthReportRequest {
  repoRoot: string;
  topic?: string;
  staleSeconds?: number;
  stdinSlugs?: string[];
}

export interface WikiHealthReport {
  orphans: { slug: string }[];
  stale: { slug: string; days_since_update: number }[];
  dead_refs: { slug: string; path: string }[];
  broken_links: { source_slug: string; target_slug: string }[];
  broken_xwiki: {
    source_slug: string;
    target_wiki: string;
    target_slug: string;
  }[];
  missing_sources: { slug: string; source_id: string }[];
  unused_sources: { slug: string; source_id: string }[];
  legacy_frontmatter: { slug: string; fields: string[] }[];
  unfixable_sources: { slug: string; source: string }[];
  duplicate_sources: { slug: string; source_id: string }[];
  empty_topics: { slug: string }[];
  empty_pages: { slug: string }[];
  slug_collisions: { slug: string; paths: string[] }[];
}

export type CollectWikiHealthReport = (
  request: CollectWikiHealthReportRequest,
) => Promise<WikiHealthReport>;

export { DEFAULT_STALE_SECONDS };

export async function checkWikiHealth(
  request: CheckWikiHealthRequest,
): Promise<WikiHealthReport> {
  const repoRoot = await resolveWikiRoot({
    cwd: request.cwd,
    wiki: request.wiki,
  });
  return collectWikiHealthReport({
    repoRoot,
    topic: request.topic,
    staleSeconds: request.staleSeconds,
    stdinSlugs: request.stdinSlugs,
  });
}

export async function collectWikiHealthReport(
  request: CollectWikiHealthReportRequest,
): Promise<WikiHealthReport> {
  return wikiHealthReportFromIndexerReport(
    await collectIndexerHealthReport({
      repoRoot: request.repoRoot,
      topic: request.topic,
      staleSeconds: request.staleSeconds,
      stdinSlugs: request.stdinSlugs,
    }),
  );
}

function wikiHealthReportFromIndexerReport(
  report: HealthReport,
): WikiHealthReport {
  return {
    orphans: report.orphans.map((item) => ({ slug: item.slug })),
    stale: report.stale.map((item) => ({
      slug: item.slug,
      days_since_update: item.days_since_update,
    })),
    dead_refs: report.dead_refs.map((item) => ({
      slug: item.slug,
      path: item.path,
    })),
    broken_links: report.broken_links.map((item) => ({
      source_slug: item.source_slug,
      target_slug: item.target_slug,
    })),
    broken_xwiki: report.broken_xwiki.map((item) => ({
      source_slug: item.source_slug,
      target_wiki: item.target_wiki,
      target_slug: item.target_slug,
    })),
    missing_sources: report.missing_sources.map((item) => ({
      slug: item.slug,
      source_id: item.source_id,
    })),
    unused_sources: report.unused_sources.map((item) => ({
      slug: item.slug,
      source_id: item.source_id,
    })),
    legacy_frontmatter: report.legacy_frontmatter.map((item) => ({
      slug: item.slug,
      fields: [...item.fields],
    })),
    unfixable_sources: report.unfixable_sources.map((item) => ({
      slug: item.slug,
      source: item.source,
    })),
    duplicate_sources: report.duplicate_sources.map((item) => ({
      slug: item.slug,
      source_id: item.source_id,
    })),
    empty_topics: report.empty_topics.map((item) => ({ slug: item.slug })),
    empty_pages: report.empty_pages.map((item) => ({ slug: item.slug })),
    slug_collisions: report.slug_collisions.map((item) => ({
      slug: item.slug,
      paths: [...item.paths],
    })),
  };
}

File: src/codealmanac/cli/render/health.py
from codealmanac.cli.render.common import print_json_model
from codealmanac.cli.render.style import style
from codealmanac.services.health.models import ValidationIssue, ValidationResult
from codealmanac.services.index.models import HealthReport


def render_health(report: HealthReport, json_output: bool) -> None:
    if json_output:
        print_json_model(report)
        return
    blue = style.BLUE
    dim = style.DIM
    rst = style.RST
    sections = [
        ("orphans", tuple(f"{blue}{item.slug}{rst}" for item in report.orphans)),
        (
            "dead-refs",
            tuple(
                f"{blue}{item.slug}{rst}  references {item.path} {dim}(missing){rst}"
                for item in report.dead_refs
            ),
        ),
        (
            "broken-links",
            tuple(
                f"{blue}{item.source_slug}{rst} -> {item.target_slug} "
                f"{dim}(target does not exist){rst}"
                for item in report.broken_links
            ),
        ),
        (
            "broken-xwiki",
            tuple(
                f"{blue}{item.source_slug}{rst} -> "
                f"{item.target_wiki}:{item.target_slug} "
                f"{dim}(wiki unregistered or unreachable){rst}"
                for item in report.broken_xwiki
            ),
        ),
        (
            "empty-topics",
            tuple(f"{blue}{item.slug}{rst}" for item in report.empty_topics),
        ),
        (
            "empty-pages",
            tuple(f"{blue}{item.slug}{rst}" for item in report.empty_pages),
        ),
        (
            "missing-source-citations",
            tuple(
                f"{blue}{item.slug}{rst} cites {item.source_id} "
                f"{dim}(missing source){rst}"
                for item in report.missing_source_citations
            ),
        ),
        (
            "unused-sources",
            tuple(
                f"{blue}{item.slug}{rst} lists {item.source_id} {dim}(not cited){rst}"
                for item in report.unused_sources
            ),
        ),
        (
            "duplicate-sources",
            tuple(
                f"{blue}{item.slug}{rst} repeats {item.source_id}"
                for item in report.duplicate_sources
            ),
        ),
    ]
    rendered = [health_section(name, rows) for name, rows in sections]
    print("\n\n".join(rendered))


def health_section(name: str, rows: tuple[str, ...]) -> str:
    if not rows:
        return f"{style.BOLD}{name}{style.RST} {style.GREEN}(0): ok{style.RST}"
    lines = [f"{style.BOLD}{name}{style.RST} {style.RED}({len(rows)}){style.RST}:"]
    lines.extend(f"  {row}" for row in rows)
    return "\n".join(lines)


def render_validate(result: ValidationResult, json_output: bool) -> None:
    if json_output:
        print_json_model(result)
        return
    if result.ok:
        status = f"{style.GREEN}ok{style.RST}"
    else:
        status = f"{style.RED}failed{style.RST}"
    print(f"{style.BOLD}validate{style.RST}: {status}")
    print(f"{style.DIM}wiki:{style.RST}  {result.repository_name}")
    print(f"{style.DIM}path:{style.RST}  {result.almanac_path}")
    if result.index is not None:
        print(f"{style.DIM}index:{style.RST} {result.index.pages_indexed} pages")
    if result.ok:
        return
    print(f"issues ({len(result.issues)}):")
    for issue in result.issues:
        print(f"  {format_validation_issue(issue)}")


def format_validation_issue(issue: ValidationIssue) -> str:
    location = issue.page or issue.path or ""
    subject = f"{style.BLUE}{location}{style.RST} " if location else ""
    return f"{subject}{issue.message} {style.DIM}({issue.category}){style.RST}"
